Write timestamp spike

tcharding opened the pull request:
As part of #1044 write a spike about timestamps.

tcharding commented:
@comit-network/rust-devs please be super nit-picky when reviewing this one.  Its my first spike, I'd like to get all the small details ironed out quickly as possible.  #mercylessreview Thanks.

thomaseizinger reviewed (dismissed):
Looks mostly fine to me. What I would like to see though is a recommendation of the author, which events we should now timestamp and whether those timestamps are the ones from the blockchain or when we received them.

Resolve this ambiguity and we are good to go!

tcharding commented:
> Looks mostly fine to me. What I would like to see though is a recommendation of the author

I did not include an opinion in this spike because I do not understand the motivation for timestamps.  The issue states only 'This just came up in a discussion: Events should have timestamps'.  Discussion meeting scheduled.

D4nte commented:
The aim of the timestamp is for the client to have an idea of the order of events and have the information. This would allow comit-i to display the events in the right order or it could allow a trading bot to analyse how much time a given peer is taking to do his part of a swap (ie, if a given peer is always taking a long time then he is likely to be using the option and hence the trading bot may decide to take action against this specific peer).

There is no possible "spoofing" in this scenario as we currently always assume that a client can trust its COMIT node.

As a base assumption for this spike, I would not expect comit_node to get any timestamp from another COMIT node. As you said, the comit_node should either make a timestamp by registering when a message arrives/is sent (from another COMIT node) or looking at the blockchain messages.

For this spike I'd expect to have the list of events, (swap request sent, acceptance of swap received, decline of swap received, transaction on blockchain seen/etc) and for each event have a list of the possible timestamps (ie, timestamp of when a message was sent vs timestamp of the transaction vs timestamp of the block etc) with a declaration of the risk for each (e.g., using timestamp of the block gives a precision of 10min etc).

tcharding commented:
Nice description @D4nte, thanks.  Now I better understand the spike.  I have cancelled the discussion meeting.

tcharding committed:
Add timestamp spike

D4nte commented on 0010-timestamps.adoc:
```diff
@@ -0,0 +1,78 @@
+=== Events from Alice's perspective
+
+==== happy swap
+- swap accept received (fund available)
```
I'd add `swap request sent`

tcharding replied:
Thank you, added.

D4nte commented on 0010-timestamps.adoc:
```diff
@@ -0,0 +1,78 @@
+- swap accept sent
+- Alpha ledger funded (fund available)
+- Beta ledger funded (refund available)
+- Beta asset redeemed (by Alice, redeem available)
```
I'd add `beta/alpha asset refunded` for both Alice and Bob.

da-kami replied:
Additionally I'd suggest renaming `happy swap` to something else that also depicts refund (e.g. `accepted swap`).
(`happy swap` sounds like the success scenario to me, which does not include refund from my point of view.)

tcharding replied:
I intentionally left these out (hence name 'happy path').  Adding a section to the document stating the reasons.  You guys can ack/nack from there :)

tcharding replied:
Rethought the above comment.  Added `refund`.

D4nte commented on 0010-timestamps.adoc:
```diff
@@ -0,0 +1,78 @@
+
+== Suggestions
+
+I suggest we timestamp the events above using seconds since epoch.
```
So if I understand correctly, your suggestion is to only timestamp the event from the moment `cnd` is made aware of them and ignore any timestamp coming from the blockchain.

It sounds reasonable but did you consider using timestamps from the blockchain?
E.g, when a `fund` transaction is seed, instead of saving the timestamp of when cnd is made aware, use the timestamp of the blockchain transaction (which is actually the timestamp of the block in which the transaction was confirmed).

Also, can we think about what would happen if we were to consider 6 confirmations for the `fund` event to be considered done?

We would have to assume that btsieve would return the transaction once all confirmations are done. So 1 hour after the transaction is broadcasted. Is that fine?

da-kami replied:
Question: What do we want to achieve with the timestamp?
If the sole purpose is, to put the event in order (so that comit-i can display them in the correct sequence) I think a timestamp by the comit-node would be sufficient. 

If we want to depict block confirmations in the timestamp then this would actually be a different timestamp for me. 
Do we have other use-cases besides the event order in comit-i that involve the timestamp?

tcharding replied:
I limited the scope of these timestamps to just being of use to clients i.e. comit-i (stated in the document).  I do not think we should bother with blockchain times.

D4nte replied:
> I limited the scope of these timestamps to just being of use to clients i.e. comit-i (stated in the document). I do not think we should bother with blockchain times.

Can you add that to the spike then please?

tcharding replied:
Added 'I suggest that all timestamps are created by the cnd since they are only useful to the cnd I feel it adds no real benefit to get verified times from the blockchain.'

tcharding replied:
Did patch with a this added as a 'suggestion'.

D4nte replied:
Getting timestamp from the blockchain is not all about getting **verified** times, it's also about getting all events especially if we implement the confirmation feature.

Please note that I agree with your proposal, I just want to not leave any stones unturned.

tcharding commented:
> So if I understand correctly, your suggestion is to only timestamp the event from the moment cnd is made aware of them and ignore any timestamp coming from the blockchain.

Yes

tcharding committed:
Add refund events

It was suggested that we include 'refund' events.  Good idea.  Add refund events
to the list of events.  Restructure the list into a single list.  Use
highlighting to show events from a different sequence flow.

tcharding committed:
Add suggestion on using cnd to create all timestamps

tcharding committed:
Add events for ledger and cnd

D4nte commented on 0010-timestamps.adoc:
```diff
@@ -0,0 +1,73 @@
+These events are marked below with '(ledger + cnd)', basically these are all the events except the initial COMIT messages.
+
+- swap request sent/received
+- --- swap decline sent/received
```
what does `---` mean?

tcharding replied:
Damn, I was hoping that would be self evident.  it signals that all the other events form a sequence and the '---' events are branches.

D4nte replied:
huum, maybe find a better way?

D4nte commented on 0010-timestamps.adoc:
```diff
@@ -0,0 +1,73 @@
+== Conclusion
+
+In conclusion I recommend adding a timestamp generated by the cnd for each of the message events.
+For each notification from btsieve to cnd, recommend passing timestamp information (from the ledger) from btsieve to cnd and storing this information in cnd to make available to comt-i.
```
```suggestion
For each notification from btsieve to cnd, recommend passing timestamp information (from the ledger) from btsieve to cnd and storing this information in cnd to make available to comit-i.
```

D4nte commented on 0010-timestamps.adoc:
```diff
@@ -0,0 +1,73 @@
+
+In conclusion I recommend adding a timestamp generated by the cnd for each of the message events.
+For each notification from btsieve to cnd, recommend passing timestamp information (from the ledger) from btsieve to cnd and storing this information in cnd to make available to comt-i.
+All timestamps should be made available to comit-i.
```
So if I understand correctly you suggest we get the blockchain timestamp + cnd timestamp for all blockchain events?

tcharding replied:
Yes

D4nte replied:
Could you make it more explicit?

da-kami replied:
```suggestion
Both the cnd timestamps and the ledger timestamp should be made available to comit-i.
```
If this is what you mean I'd phrase it like this (or similar).

Commit:
Update 0010-timestamps.adoc

Co-Authored-By: Franck Royer <franck@coblox.tech>

D4nte reviewed (approved):
Approved, it'd be great if you could sort out the two cosmetic comments.

tcharding commented:
> Approved, it'd be great if you could sort out the two cosmetic comments.

The '---'?   Whats the other please @D4nte?

D4nte commented:
> > Approved, it'd be great if you could sort out the two cosmetic comments.
> 
> The '---'? Whats the other please @D4nte?

Describe your recommendation in a more explicit manner.

da-kami reviewed (approved):
Approved, but added one last nitpicky suggestion :)

Commit:
Update 0010-timestamps.adoc

Co-Authored-By: Daniel Karzel <D.Karzel@gmail.com>

tcharding committed:
Make recomendations more explicit
